Attempts/Questions klonen

tumidi opened the pull request:
- Backend
  - Implementiert API-Endpunkte für's Klonen.
- Frontend
  - Fügt die entsprechenden Actions den Buttons hinzu.
  - Nach dem Klonen springt die Ansicht auf die Übersichtsseite und hebt geklonten Eintrag farblich hervor.
  
  Issue #238

tumidi committed:
feat(webserver): implement attempt cloning

Ref: #238

tumidi committed:
feat(frontend): add attempt cloning

Ref: #238

tumidi committed:
feat(webserver): implement question cloning

Ref: #238

tumidi committed:
feat(frontend): add question cloning

Also, refactor scrolling/highlighting behavior into `useHighlightOnInsert` composable.

Closes: #238

tumidi committed:
chore(tests): prevent pytest from recursing into `frontend` folder

pytest tries to collect `*.py` files in `frontend/node_modules`, which causes errors.

MHajoha commented on frontend/src/composables/common/useHighlightOnInsert.ts:
```diff
@@ -0,0 +1,88 @@
+        },
+        highlightedIds,
+    }
+}
```
Ich verstehe nicht ganz die Dualität zwischen dem Watch auf `items` und dem Hinzufügen neuer Items zu `pendingIds`. (Das könnte aber auch einen Vue-Skill-Issue meinerseits sein.)

Damit etwas gehighlighted (highgelit?) wird, muss erst `handleNewItem` aufgerufen, und dann das neue Element zu items hinzugefügt werden, oder? Das wirkt relativ fragil bzgl. Race Conditions.

Und wir wollen doch eh immer das neue Element highlighten (oder nach Navigation das aus der History). Brauchen wir da überhaupt `handleNewItem`?

tumidi replied:
Es kann erst zum neuen Eintrag gescrollt (und geheilei... hervorgehoben) werden, wenn die Daten gerefetched und gerendert sind.

Die aktuelle Implementierung nimmt zu behandelnde IDs auf zwei Arten an:
1. über `handleNewItem`
2. über einen konfigurierbaren key in `history.state`, um die Mechanik auch nach dem Navigieren zu unterstützen.

`pendingIds` sind die IDs, die in der Zukunft noch behandelt werden müssen.

`items` ist ein (reaktives) ID-indiziertes Objekt von Items, die das composable überwacht und reagiert, sobald eines der `pendingIds` darin auftaucht. In unserem Fall kommen die `items` direkt aus dem Pinia Colada Store, also von der API.

> und dann das neue Element zu items hinzugefügt werden, oder? Das wirkt relativ fragil bzgl. Race Conditions.

Diese Probleme sehe ich:

- dass wir mehrere Clones gleichzeitig haben, und die `scrollTos` quasi miteinander kämpfen (passiert wenn man wild auf clone klickt).
- aus irgendeinem Grund das neue Item in `items` ankommt, bevor noch das Composable das erste Mal läuft, also die Logik in `watch` nie getriggert wird (könnte prinzipiell passieren, z.B. nach einer async Navigation).

Ich bereite gerade einen Patch vor, der

- auf einem globalen Store basiert, der "Operations" trackt.
  - So kann ich global alle Buttons deaktivieren, so dass immer nur eine Clone-Operation zur selben Zeit läuft.
  - Der State überlebt auch eine Navigation. Es muss kein State mehr in `history.state` mehr versteckt werden.
  - `handleNewItem` wird überflüssig.
- neue Item auch erkennt, wenn das Composable läuft nachdem das Item ankommt.
- das Composable wird dadurch etwas einfacher und hat weniger Verantwortlichkeiten.

tumidi replied:
Das vorherige Composable `useHighlightOnInsert` habe ich in drei unabhängige Komponenten mit klaren Verantwortlichkeiten aufgeteilt.

fccc889c871329355b52bb2a7f6c41668f4f019d

### `pendingOperations` store

Der neue Patch führt einen Store ein, der State für asynchrone Operationen zentral trackt.

- Damit wird das State-Management vereinheitlicht, insbesondere
  - keine Parent-Child-Kommunikation über Events und `handleNewItem` mehr, und
  - keine Kommunikation durch `history.state`.
- Es wird nur noch eine Operationen gleichzeitig erlaubt. (Unterbindet die gleichzeitige Ausführung mehrerer `scrollTo`s).
- Während Operationen sind Aktions-Buttons und auch Page-Transitions deaktiviert, was verhindert, dass das UI in einen inkonistenten State gelangt, zB dass die Person von der Seite wegnavigiert bevor `deferred-item` konsumiert wurde, etc.

### `useDeferredItem`

- Überwacht eine Collection und reagiert wenn ein erwartetes Item erscheint.
- Konsumiert dafür ein `deferred-item` und ruft einen beliebigen Handler auf.
- Im Gegensatz zu vorher is nun auch das andere Timing-Issue gelöst, nämlich, dass ein ein deferred Item übersehen wird, weil das Composable später ausgeführt wird als das Item tatsächlich ankommt.

### `useHintItem`

Die ganze DOM-related `scrollTo`/Highlight-Logik.

- Nutzt zusätzlich einen Timer als Fallback, falls `animationend` aus irgendeinem Grund nicht feuert, so dass das UI wieder freigegeben wird.
- Verzichtet auf `window.getElementById`, sondern setzt stattdessen auf eine `ref`-Assigner-Funktion, um an die DOM-Elemente zu gelangen, was Vue-idiomatischer ist.

MHajoha commented on frontend/src/queries/attempt.ts:
```diff
@@ -98,18 +100,36 @@ function usePostAttemptMutation(questionId: string, attemptId: string) {
+ * @returns An mutation return object.
+ */
+function usePostAttemptCloneMutation(questionId: string, attemptId: string, newAttemptId: string) {
+    const invalidateAttempt = useInvalidateAttempt(questionId, attemptId)
```
Der geklonte Attempt müsste doch gleich bleiben, warum invalidieren?

tumidi replied:
Ja, richtig, da sollte eigentlich nur die Attempt-Liste invalidiert werden.

4fceb080e4574ca5f2f3966150398a6e6b6003f1

MHajoha commented on questionpy_sdk/webserver/controllers/attempt/controller.py:
```diff
@@ -99,6 +99,31 @@ async def score_attempt(self, question_id: str, attempt_id: str) -> None:
 
         await self._state_manager.write_attempt_score(question_id, attempt_id, attempt_scored)
 
+    async def clone_attempt(self, question_id: str, attempt_id: str, new_attempt_id: str) -> None:
```
Ich hätte hier tatsächlich einfach `cp -r` (bzw. `shutil.copytree`) gemacht. Spricht da etwas gegen?

tumidi replied:
Du meinst in `AttemptController.clone_attempt()`? Dort ginge das im aktuellen Design gar nicht. Ich hatte die Controller und das state management mal decoupled irgendwann.
Die Controller reden nur mir dem `StateManager`-Protokoll, kennen gar keine Ordner oder Dateien. Dadurch könnten wir später mal das state management relativ einfach durch irgendwas anderes ersetzen, SQLite oder so.

Ich glaube, die eigenliche Schwäche am aktuellen Design ist, dass der Webserver keinen Data Layer besitzt und diese Logik auf die Controllermethoden abwälzt. Da gibt's auf jeden Fall Refaktorisierungspotential, um die Controllermethoden schlanker und übersichtlicher zu machen.

Davon abgesehen würde `cp -r` nicht erkennen, ob ein Attempt überhaupt gültig ist, also zumindest valide data- und score-Dateien besitzt, und natürlich auch einen 626 GiB großen tarball mitkopieren, den jemand scherzeshalber oder aus Versehen im Attempt-Ordner lagert.

MHajoha replied:
Ja, ich meinte schon das Hinzufügen einer `StateManager.clone_attempt()` Methode. Bzgl. Data Layer: Bitte nicht noch eine Abstraktionsebene, die Trennung lohnt sich hier nicht.

Ich glaube nicht, dass uns über 100% PEBKAC-Fälle wie das Tarball-Beispiel sonderlich viele Gedanken machen sollten.

MHajoha commented on frontend/src/composables/question/useCloneQuestion.ts:
```diff
@@ -0,0 +1,31 @@
+ * @returns The ID of the new question.
+ */
+function useCloneQuestion(questionId: string) {
+    const newQuestionId = generateId()
```
Ah, das war mir gar nicht aufgefallen, aber Jan hatte dadurch einen Bug: Wenn die nutzende Komponente in der Zwischenzeit nicht neu gerendert wird, wird hier versucht, die gleiche ID nochmal zu verwenden. Dann wird leise die geklonte Frage nochmal überschrieben. 

Vielleicht wäre da auch ein Check im Backend cool, dann wäre das früher aufgefallen.

tumidi replied:
Good catch!

9b1c6875d6c6573fc94c7845f0f8d2e4f6e7ae9f fixt das im Frontend

f5bb9afde64b253ff870fec387c445896ef6d0ef gibt 409 bei schon existierender ID, außerdem ist mir aufgefallen, dass ich bei Attempt gar nicht 404 korrekt berücksichtigt habe.

MHajoha commented on frontend/src/composables/attempt/useCloneAttempt.ts:
```diff
@@ -0,0 +1,33 @@
+ * @returns The ID of the new attempt.
+ */
+function useCloneAttempt(questionId: string, attemptId: string) {
+    const newAttemptId = generateId()
```
Same here

tumidi replied:
siehe https://github.com/questionpy-org/questionpy-sdk/pull/240#discussion_r2468906501

tumidi committed:
fix(frontend): generate a fresh ID for each clone invocation

tumidi committed:
fix(backend): handle `HTTPNotFound` and `HTTPConflict` consistently in clone operations

tumidi committed:
fix(frontend): avoid unnecessary attempt invalidation when cloning

tumidi committed:
feat(frontend): add `pendingOperations` store

Split `useHighlightOnInsert` into focused composables with well-defined responsibilities.

tumidi committed:
feat(frontend): UI improvements, refactors, and style updates

- Add `<BackLink>` component.
- Implement `<LoadingIndicator>` component.
- Switch to Pinia Colada's `isPending` (tracks first request) instead of `asyncStatus`.
- Replace manual margins with Bootstrap `vstack` in main page and list views

tumidi committed:
refactor(frontend): extract `<ListView>` from `<AttemptList>`, `<QuestionList>`

tumidi commented:
Mit der Einführung von `pendingOperations` haben sich ein paar Refaktorisierungen aufgedrängt, die ich der Übersicht halber separat committed und mit ein paar anderen kleineren Änderungen, die ich eh noch in der Pipeline hatte in a54a5e93abee389aaa7ad10c96c099f8c5f3e1ad zusammengeworfen habe.

In 9d374cbadb272de8afc8452a9c466ef900e19305 habe ich schlussendlich noch eine `<ListView>` extrahiert, da die duplizierte Logik in den Komponenten jetzt gefühlt einen Schwellwert erreicht hatte.
